[Courses App] ADD: course list view 추가

tonic523 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
courses app의 course list api 구현
		 
## 수정 사항들 자세한 내용
- course list 구현
- category 분류 list 구현
- sub_category 분류 list 구현
- course list 정렬 구현
- Rest 아키텍처에 맞춰 url 구현
- unit test 구현
- invalid_value error 구현
- pagination 구현
- 검색 기능 구현


```json
{
courses: [
{
id: 1,
title: "축구계의 끝판왕",
sub_category: "스포츠",
user: "신승호",
like: 1,
price: "30000.00",
thumbnail: "null.jpg"
},
{
id: 2,
title: "농구계의 끝판왕",
sub_category: "스포츠",
user: "홍연우",
like: 1,
price: "40000.00",
thumbnail: "null.jpg"
},
```
## unit test 결과
test 함수
- 카테고리 별 리스트 (SUCCESS)
- 카테고리 별 리스트 (INVALID_VALUE)
- 세부 카테고리 별 리스트 (SUCCESS)
- 세부 카테고리 별 리스트 (INVALID_VALUE)
- 리스트 정렬 (SUCCESS)
- 리스트 정렬 (INVALID_VALUE)
- pagination (SUCCESS)
- pagination (INVALID_VALUE)

## ![스크린샷, 2021-05-31 20-30-45](https://user-images.githubusercontent.com/59171113/120186986-327a1a80-c24f-11eb-9b2f-edf8c0d905cd.png)

## 기타 질문 및 특이 사항

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.
- [x] 유닛 테스트를 구현 하였고 모든 유닛 테스트가 통과 된것을 확인 하였습니다.

kst6294 commented on courses/urls.py:
```diff
@@ -0,0 +1,9 @@
+from courses.views import ListView
+
+urlpatterns = [
+    path('', ListView.as_view()),
+    path('/<str:category>', ListView.as_view()),
+    path('/<str:category>/<str:sub_category>', ListView.as_view())
```
프론트로 받아오는 `category` 와 `sub_category` 에 해당하는 course 의 목록을 `조회` 하는 것이기 때문에 path parameter 보다는 query parameter 가 더 적절할 것 같습니다.

그리고 데이터를 조회할 때, 중복값이 발생할 확률이 있는 category의 name 보다는 pk 값인 id 를 받아 오는 것이 더 적절해 보입니다. 

정리하자면, 현재와 같은 패스파라미터 형식으로 작성을 꼭 해야 한다면, 아래와 같게 되겠지요
```
path('/category/<int:category_id>/sub-category/<int:sub_category_id>, ListView.as_view())
```
결과 >>> 8000/course/category/1/sub-category/1 

어떠한 방법이 더 적절한 것 같으신가요?

tonic523 replied:
query parameter로 하는게 더 직관적이여서 좋은 것 같아요!
카테고리 밑에 세부 카테고리가 있다고 생각해서 위처럼 작성했는데 생각해보니 세부 카테고리만 있어도 조회가 가능하기 때문에
패스파라미터 형식은 불필요한 내용까지 넣는 것 같아요
수정해서 다시 올리겠습니다

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,49 @@
+                    )
+        # 정렬
+        sort_name = request.GET.get('sort', None)
+        if sort_name == 'latest':
```
sort 에 관한 내용도 정규화 혹은 상수로 처리해주세요!

tonic523 replied:
예를 들어
LATEST = 'latest' 처럼 상수로 만들면 되나요?
아니면 조건문을 수정해야 되나요?

kst6294 replied:
LATEST = 1
SATISFACTION = 2
LIKES = 3

이런식으로 정리해주시면, query parameter 에서도 course?sort=1 이런식으로 할 수 있겠네요!
```
if sort == LIKES:
    ~~~
```

tonic523 replied:
오 인덱스로 해주는 방법도 있네요!
아니면 parameter값이 key, order_by에 들어가는 값이 value인 dictionary를 만들어서 사용하면 어떤가요?
```python
sort_name = request.GET.get('sort', None)
if sort_name:
    my_dict = {
        'lastest'     : '-created_at',
        'satisfaction': '-review_count',
        'likes'       : '-like_count'
    }
    course_list = course_list.annotate(review_count = Count('review'))
    course_list = course_list.annotate(like_count = Count('liked_user'))
    course_list.order_by(my_dict[sort_name])
```

kst6294 replied:
이 방법으로 하셔도 괜찮을 것 같습니다!! 👍

kst6294 reviewed (commented):
피드백 확인해주시고, 의견 부탁드립니다!

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,48 @@
+        course_list     = []
+        # 카테고리 분류
+        if category_id:
+            course_list = Course.objects.filter(
+                sub_category__in = SubCategory.objects.filter(
+                    category = Category.objects.get(id=category_id)
```
```suggestion
            course_list = Course.objects.filter(sub_category__category_id = category_id
```

이런식으로 줄여서 가독성을 높일 수 있겠네요!

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,48 @@
+                    )
+                )
+        if sub_category_id:
+            course_list = Course.objects.filter(
+                sub_category = SubCategory.objects.get(id=sub_category_id)
```
여기도 위의 피드백 참고하셔서 수정해주시면 좋겠습니다!

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,48 @@
+        sub_category_id = request.GET.get('sub_category', None)
+        course_list     = []
+        # 카테고리 분류
+        if category_id:
```
혹시 category_id 와 sub_category_id 두개가 동시에 들어오는 경우에는 어떻게 처리를 하시나요?

tonic523 replied:
sub_category로 분류하는 filter를 후순위로 두어서 두개가 동시에 들어올 때는 sub_category로 분류하게 처리했습니다.

tonic523 replied:
위 방법이면 둘 다 들어왔을 때 sub_category로만 분류해야 되는데 category 분류도 하게 되네요

tonic523 replied:
```python
        if sub_category_id:
            course_list = Course.objects.filter(sub_category__id = sub_category_id)
        elif category_id:
            course_list = Course.objects.filter(sub_category__category_id = category_id)
        else:
            course_list = Course.objects.all()
```

tonic523 replied:
이 코드로 바꾸면 sub_category가 들어왔을 때 불필요한 분류를 최소화할 수 있을 것 같아요

kst6294 replied:
둘 값을 모두 받았을 때, category 를 통해서만 조회를 하는 방향을 의도하신게 맞는거죠? 👍

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,48 @@
+    def get(self, request):
+        category_id     = request.GET.get('category', None)
+        sub_category_id = request.GET.get('sub_category', None)
+        course_list     = []
```
불필요한 리스트 선언도 제거할 수 있겠네요!

tonic523 replied:
두 값을 모두 받았을 때 sub_category 를 통해서만 조회하는 방향으로 하려고 합니다!

kst6294 reviewed (commented):
수정사항 마저 반영해주시고, 수정된 코드를 통한 통신 내역도 같이 업데이트 해주세요 👍

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,43 @@
+                'satisfaction': '-review_count',
+                'likes'       : '-like_count'
+            }
+            course_list = course_list.annotate(review_count = Count('review'))
```
review_count 는 사용하지 않나요?

tonic523 replied:
sort 값이 들어오면 무조건 실행하게 했는데 조건절로 묶어서 하는게 더 좋은 방법인 것 같아요

kst6294 commented on courses/views.py:
```diff
@@ -1,3 +1,43 @@
+                'title'       : course.title,
+                'sub_category': course.sub_category.name,
+                'user'        : course.user.korean_name,
+                'like'        : course.liked_user.count(),
```
```suggestion
                'like'        : course.liked_user,
```

이렇게 확인해보시겠어요?
이미 count 를 진행했기 때문에 다시 count() 를 하지 않아도 괜찮아보입니다.

tonic523 replied:
아 넵 annotate로 만든 주석 필드(like_count)로 바꾸겠습니다.

kst6294 reviewed (commented):
수정바랍니다!

kst6294 commented:
unittest 만 적용해주세요 👍

tonic523 commented:
순태님 list api에서 전체 리스트, 카테고리 별 리스트, 정렬 이 3가지 기능이 있는데 각 기능 별로 따로 test를 만드나요?
한 test 객체안에 3가지의 함수 test를 만들어야 되는지 test객체를 3가지를 만들어야 되는지 헷갈리네요 ㅠㅠ
제 생각은 기능들이 거의 똑같은 testcase를  사용해서 3가지의 함수를 만드는게 좋은 것 같습니다.

kst6294 commented on courses/tests.py:
```diff
@@ -47,4 +50,121 @@ def test_category_get_view(self):
+                
+    def tearDown(self):
+        Category.objects.all().delete()
+        SubCategory.objects.all().delete()
```
User, Target 도 삭제해야할 것 같습니다~

kst6294 commented on courses/tests.py:
```diff
@@ -47,4 +50,121 @@ def test_category_get_view(self):
+    def test_all_list(self):
+        client     = Client()
+        response   = client.get('/courses')
+        course_obj = Course.objects.all().annotate(like_count = Count('liked_user'))
```
이 코드는 `/courses` 엔드포인트에 요청을 보냈을 때 해당하는 그 뷰 안에 이미 존재하는 코드이기 때문에, 굳이 테스트코드 안에 작성하지 않으셔도 괜찮습니다. 결국 엔드포인트에 요청하는 방식에 따라 실제 뷰 안에 있는 로직이 돌면서 그 응답만을 테스트 하게 됩니다.

kst6294 commented on courses/tests.py:
```diff
@@ -47,4 +50,121 @@ def test_category_get_view(self):
+        })
+        self.assertEqual(response.status_code, 200)
+
+    def test_list_by_category(self):
```
그리고 테스트 코드는 단순히 `성공` 만을 위한 코드 보다는 여러가지 실패하는 환경도 같이 테스트하게 됩니다.
예를들면 포스트 방식으로 테스트를 한다고 가정했을 때, 바디에 데이터가 없는 경우, 키 값이 없는 경우 등에 대한 실패도 고려해서 작성한다는 것이 되겠죠.

tonic523 replied:
넵! 카테고리, 세부 카테고리, 정렬할 때 없는 키값이 들어갈 때 에러처리했습니다.

kst6294 reviewed (commented):
테스트코드 잘 작성하고 계십니다! 
수정사항 확인해주세요 👍

tonic523 commented:
순태님 제가 paginator 를 추가했는데 2가지 방법을 사용했습니다.
1. offset, limit 를 지정해서 python문으로 사용하는 방법
2. django에서 제공하는 Paginator를 사용하는 방법
page를 쿼리파라미터로 받아서 객체 리스트만 주면 돼서 굳이 Paginator를 사용하지 않아도 될 것 같은데 어떤 방법이 좋은가요?

tonic523 commented:
검색 기능 추가했습니다

kst6294 commented on courses/tests.py:
```diff
@@ -47,4 +51,152 @@ def test_category_get_view(self):
+    def test_list_by_category_success(self):
+        client      = Client()
+        response    = client.get('/courses?sub_category=4')
+        course_list = Course.objects.filter(sub_category = 4)
+        course_obj  = course_list.annotate(like_count = Count('liked_user'))
```
이미 view 로직 상에 있는 코드를 다시 작성하지 않으셔도 괜찮습니다.
테스트가 동작하면 view 에 있는 코드도 동작하게 됩니다!

tonic523 replied:
```python
        course_list = Course.objects.filter(sub_category = 4)
        # course_obj  = course_list.annotate(like_count = Count('liked_user'))
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), {
            'courses': [
                {
                    'id'          : course.id,
                    'title'       : course.title,
                    'sub_category': course.sub_category.name,
                    'user'        : course.user.korean_name,
                    'like'        : course.like_count,
                    'price'       : '{:.2f}'.format(course.price),
                    'thumbnail'   : course.thumbnail,
                    'month'       : course.month
                }
                for course in course_list]
        })
```
제가 테스트하기 위한 객체 리스트를 만들기 위해 사용해서 위 코드를 없애면 attributeError 가 발생합니다ㅠㅠ
다른 방법으로 test를 작성해야하나요?

kst6294 replied:
한번 코드 같이 보시죠~!

kst6294 replied:
혹시 어떠한 attribute Error 인지는 확인이 안되시나요?

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +15,75 @@ def get(self, request):
+
+class ListView(View):
+    def get(self, request):
+        query = request.GET.get('query', None)
```
query 가 검색하는 키워드 인가요?
그렇다면 query 보다는 search 혹은 keyword 등의 네이밍이 더 적절할 것 같습니다.

tonic523 replied:
넵 search로 수정하겠습니다

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +15,75 @@ def get(self, request):
+        page = request.GET.get('page', None)
+        if page:
+            PAGE_SIZE = 10
+            paginator = Paginator(course_list, PAGE_SIZE)
+            list_by_page = paginator.page(page)
+            course_list = list_by_page.object_list
+            # limit       = PAGE_SIZE * int(page)
+            # offset      = limit - PAGE_SIZE
+            # course_list = course_list[offset:limit]
```
1. offset, limit 를 지정해서 python문으로 사용하는 방법
2. django에서 제공하는 Paginator를 사용하는 방법

질문에 대한 답변으로는, limit 과 offset 을 어떻게 사용하는지 알고 계신다면, 굳이 어떤 방법이 더 좋은지 구분하기 보다는, 사용하기 편한 방법을 채택하는 게 좋을 듯 합니다 .

limit과 offset을 통해서 어떻게 pagination 을 구현할 수 있는지 말씀해주실 수 있을까요?

tonic523 replied:
query parameter로 page 값을 받습니다. 예를 들어 page=2 가 들어오면 `limit=20` , `offset=10`으로 설정 됩니다.
이 값을 통해 course_list를 슬라이싱하여 반환합니다.
아 여기 마지막페이지일 때는 슬라이싱하는 코드가 바꿔야 될 것 같아요! `course_list[offset:]`

kst6294 reviewed (commented):
확인바랍니다!

kst6294 commented on courses/tests.py:
```diff
@@ -47,4 +48,210 @@ def test_category_get_view(self):
+        client   = Client()
+        response = client.get('/courses?category=2')
+        self.assertEqual(response.status_code, 200)
+        print(response.json())
```
불필요한 print 는 제외해주세요

tonic523 replied:
넵 수정하겠습니다

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +19,88 @@ def get(self, request):
+            course_list = list_by_page.object_list
+            page_list   = [page for page in paginator.page_range]
+        
+        access_token = request.headers.get("Authorization")
+        if access_token:
+            token_payload = jwt.decode(
+                access_token,
+                SECRET_KEY,
+                algorithms=ALGORITHM
+            )
+            user = User.objects.get(id=token_payload['user_id'])
+        else: user = None
```
이코드는 뭔가요? 
데코레이터 반영하고 코드 다시 수정해주세요~!

tonic523 replied:
지금 데코레이터가 토큰을 받지 못하면 에러처리를 해서 직접 작성하는 방법으로 했습니다.
에러처리가 아닌 유저에 None값을 주는 데코레이터를 하나 더 만들어서 사용하면 될까요?

kst6294 replied:
그런데 이 뷰는 사용자의 로그인이 중요하지 않은 것 같습니다.
로그인을 해야만 어떠한 코스가 있는지 보여주는게 아니라면, 굳이 token 을 검증할 필요 없을 것 같습니다!

kst6294 replied:
이러한 환경, 즉 사용자 정보가 없어도 코드는 정상적으로 동작하되, 좋아요와 같은 로그인 된 경우와 아닌 경우에 차이가 발생하는 경우를 처리하기 위해서 두가지 정도 고려해주시면 좋을 것 같습니다! 
1. 반드시 로그인이 되어야 하는 환경에서 사용할 데코레이터/ 로그인이 되지 않아도 동작은 하지만 차이가 있는 데코레이터 
2. header 의 토큰을 검증하는 코드를 모듈화 해서 사용하는 방법 (기존의 작성된 코드를 분리) 

이렇게 고민하면 좋을 것 같습니다. 1번의 이유는 데코레이터의 사용과 동일 하고, 
2번의 이유는 지금은 비록 list view 하나에서만 사용되지만, 다른 view 에서도 동일한 방식으로 사용될 수 있기 때문에, 동일한 코드를 반복적으로 사용하는 일을 피할 수 있을 것 같습니다!

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +19,88 @@ def get(self, request):
+        sub_category_id = request.GET.get('sub_category', None)
+
+        # 카테고리 분류
+        if sub_category_id:
+            if SubCategory.objects.filter(id=sub_category_id).exists():
+                course_list = Course.objects.select_related('sub_category','user').filter(sub_category__id = sub_category_id)
+            else:
+                return JsonResponse({'message' : 'INVALID_VALUE'}, status=404)
+        elif category_id:
+            if Category.objects.filter(id=category_id).exists():
+                course_list = Course.objects.select_related('sub_category','user').filter(sub_category__category_id = category_id)
+            else:
+                return JsonResponse({'message' : 'INVALID_VALUE'}, status=404)
+        else:
+            course_list = Course.objects.select_related('sub_category','user').prefetch_related('liked_user').all()
```
가독성이 떨어지네요... 
우선 if, elif 의 사용을 줄일 수 있는 방법을 고민해주시고,
자세한건 내일 같이 보면서 고민해보시죠!

tonic523 replied:
흠 그런 것 같네요 .. 그럼 sub_category, category 두 조건을 or 로 묶어서 elif를 없애는 방법은 어떤가요?

kst6294 replied:
![image](https://user-images.githubusercontent.com/44313290/120254615-30eb3980-c2c5-11eb-800b-d49d1d2433d3.png)

이런식으로도 줄일 수 있겠네요!

kst6294 replied:
코멘트상에 작성하신 코드로 반영하시면 좋을 것 같습니다~

kst6294 reviewed (commented):
테스트 코드 작성 잘 해주셨습니다!
이제 데코레이터 반영해주시고, 가독성을 높일 수 있는 방법을 더 고민해주시면 좋겠습니다 👍

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +19,88 @@ def get(self, request):
+             }for category in Category.objects.prefetch_related('sub_categorys').all()]
         return JsonResponse({'category': results}, status=200)
+
+class ListView(View):
```
```suggestion
class CourseListView(View):
```

kst6294 commented on courses/views.py:
```diff
@@ -12,5 +19,88 @@ def get(self, request):
+
+class ListView(View):
+    def get(self, request):
+        query           = request.GET.get('query', None)
```
```suggestion
        keyword           = request.GET.get('keyword', None)
```

kst6294 reviewed (commented):
확인해주세요!

tonic523 commented:
로그인을 했을 때 리스트 중에서 좋아요 누른 클래스들은 True값을 주고 아닌 클래스들은 False 값을 주기 위해선 유저의 정보가 필요하다고 생각했는데 다른 방법이 있을까요?

tonic523 commented:
이렇게 합치는 것보다 나누는게 더 가독성에 좋은가요?
```python
        if sub_category_id or category_id:
            if SubCategory.objects.filter(id=sub_category_id).exists() or Category.objects.filter(id=category_id).exists():
                course_list = Course.objects.select_related('sub_category','user').filter(Q(sub_category__id = sub_category_id) | Q(sub_category__category_id = category_id))
            return JsonResponse({'message' : 'INVALID_VALUE'}, status=404)
        else:
            course_list = Course.objects.select_related('sub_category','user').prefetch_related('liked_user').all()
```

kst6294 reviewed (commented):
확인바랍니다!

kst6294 commented:
추가적으로, 가능하면 코멘트에 대한 댓글로 소통 부탁드립니다~!
히스토리가 구분되서요!

tonic523 committed:
ADD: course list view 추가

- 전체 list 구현
- 카테고리 별 list 구현
- 정렬 구현
- unit test 구현
- invalid_value error 구현
- pagination 구현
- 검색 기능 구현
